Adopt more smart pointers in RemoteSourceBufferProxy

RupinMittal opened the pull request:
#### b7b8fa7a163976d33d36e2d31208776bafe9baa4
<pre>
Adopt more smart pointers in RemoteSourceBufferProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=279797">https://bugs.webkit.org/show_bug.cgi?id=279797</a>
<a href="https://rdar.apple.com/136114766">rdar://136114766</a>

Reviewed by Ryosuke Niwa.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
(WebCore::SourceBufferPrivateClient::InitializationSegment::AudioTrackInformation::protectedTrack const):
(WebCore::SourceBufferPrivateClient::InitializationSegment::VideoTrackInformation::protectedTrack const):
(WebCore::SourceBufferPrivateClient::InitializationSegment::TextTrackInformation::protectedTrack const):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitializationSegment):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateHighestPresentationTimestampChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDurationChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateBufferedChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidDropSample):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveRenderingError):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateEvictionDataChanged):
(WebKit::RemoteSourceBufferProxy::append):
(WebKit::RemoteSourceBufferProxy::evictCodedFrames):
(WebKit::RemoteSourceBufferProxy::shutdown):

Canonical link: <a href="https://commits.webkit.org/283822@main">https://commits.webkit.org/283822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c9a2ce7199720f7d106a42b91dcaeb241e2f0f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46823 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71492 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18581 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54621 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18372 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54059 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12446 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70511 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42988 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34521 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39661 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15744 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16939 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61625 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73191 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61503 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58409 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61562 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9322 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2942 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10253 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43705 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44891 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43446 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -101,84 +101,74 @@ Ref<MediaPromise> RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitiali
     segmentInfo.audioTracks = segment.audioTracks.map([&](auto& audioTrackInfo) {
         auto id = audioTrackInfo.track->id();
-        m_remoteMediaPlayerProxy->addRemoteAudioTrackProxy(*audioTrackInfo.track);
+        m_remoteMediaPlayerProxy->addRemoteAudioTrackProxy(Ref { *audioTrackInfo.track });
```
We should be adding `audioTrackInfo.protectedTrack()` function instead.

RupinMittal replied:
`audioTrackInfo.track` already returns a RefPtr. Should I add a `protectedTrack` function like this?
```
struct AudioTrackInformation {
     RefPtr<MediaDescription> description;
     RefPtr<AudioTrackPrivate> track;

     Ref<AudioTrackPrivate> protectedTrack() const { return *track; }
};
```
This feels a bit unsafe since we're not checking if the track is nullptr. But then again, we aren't doing that in the current version of the code either.

rniwa replied:
oh, no. Adding a version which returns RefPtr is good so like this:
```
struct AudioTrackInformation {
     RefPtr<MediaDescription> description;
     RefPtr<AudioTrackPrivate> track;

     RefPtr<AudioTrackPrivate> protectedTrack() const { return track; }
};
```

Calling `protectedTrack()` is different from accessing `track` in that it creates a temporary RefPtr which ref's `AudioTrackPrivate`.

RupinMittal replied:
Ohh ok got it. That makes more sense.

basuke commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -101,84 +101,74 @@ Ref<MediaPromise> RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitiali
 
     // We need to wait for the CP's MediaPlayerRemote to have created all the tracks
-    return m_remoteMediaPlayerProxy->commitAllTransactions()->whenSettled(RunLoop::current(), [weakThis = ThreadSafeWeakPtr { *this }, this, segmentInfo = WTFMove(segmentInfo)](auto&& result) mutable -> Ref<MediaPromise> {
+    return m_remoteMediaPlayerProxy->commitAllTransactions()->whenSettled(Ref { RunLoop::current() }, [weakThis = ThreadSafeWeakPtr { *this }, this, segmentInfo = WTFMove(segmentInfo)](auto&& result) mutable -> Ref<MediaPromise> {
```
We have RunLoop::protectedCurrent().

RupinMittal replied:
Good point! I didn't realize this was there initially--thank you. As recommended above, I'm changing `current` to be `currentSingleton` so we should need this anymore.

rniwa commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -101,84 +101,74 @@ Ref<MediaPromise> RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitiali
 
     // We need to wait for the CP's MediaPlayerRemote to have created all the tracks
-    return m_remoteMediaPlayerProxy->commitAllTransactions()->whenSettled(RunLoop::current(), [weakThis = ThreadSafeWeakPtr { *this }, this, segmentInfo = WTFMove(segmentInfo)](auto&& result) mutable -> Ref<MediaPromise> {
+    return m_remoteMediaPlayerProxy->commitAllTransactions()->whenSettled(Ref { RunLoop::current() }, [weakThis = ThreadSafeWeakPtr { *this }, this, segmentInfo = WTFMove(segmentInfo)](auto&& result) mutable -> Ref<MediaPromise> {
```
We should probably rename `RunLoop::current` to `RunLoop::currentSingleton` to suppress the warning.

RupinMittal replied:
Given Basuke's suggestion below, do we prefer using `RunLoop::protectedCurrent()` instead of renaming `current` to `currentSingleton` later on?

rniwa replied:
We can adopt protectedCurrent yeah, but we should probably get rid of this function altogether and just rename the function to `currentSingleton`. It's pretty silly to ref counting a (thread local) singleton object.

RupinMittal replied:
Got it--that makes sense. In that case, I'll revert the `RunLoop::current` changes I've made in this patch and write a patch after this that just renames the function to `currentSingleton` and use that everywhere.

basuke commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -239,13 +229,14 @@ void RemoteSourceBufferProxy::startChangingType()
 void RemoteSourceBufferProxy::removeCodedFrames(const MediaTime& start, const MediaTime& end, const MediaTime& currentTime, CompletionHandler<void()>&& completionHandler)
 {
-    m_sourceBufferPrivate->removeCodedFrames(start, end, currentTime)->whenSettled(RunLoop::current(), WTFMove(completionHandler));
+    m_sourceBufferPrivate->removeCodedFrames(start, end, currentTime)->whenSettled(Ref { RunLoop::current() }, WTFMove(completionHandler));
```
Ditto

basuke commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -316,12 +307,12 @@ void RemoteSourceBufferProxy::setAppendWindowEnd(const MediaTime& appendWindowEn
 void RemoteSourceBufferProxy::setMaximumBufferSize(size_t size, CompletionHandler<void()>&& completionHandler)
 {
-    m_sourceBufferPrivate->setMaximumBufferSize(size)->whenSettled(RunLoop::current(), WTFMove(completionHandler));
+    m_sourceBufferPrivate->setMaximumBufferSize(size)->whenSettled(Ref { RunLoop::current() }, WTFMove(completionHandler));
```
Ditto

rniwa reviewed (approved):
We should revert RunLoop::current changes for now, and rename it to `currentSingleton` in a separate PR.

basuke commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -93,92 +95,83 @@ Ref<MediaPromise> RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitiali
 {
-    auto connection = m_connectionToWebProcess.get();
-    if (!connection)
-        return;
```
I saw this pattern in your PRs which flips the if statement condition to use `if (RefPtr xxx) zzz` statement. IMO it's generally recommended to use that pattern, but in this case we want to focus on the smart pointer refactoring. If you change the meaning of condition, reviewers need to validate the rest of code. That's unnecessary work for reviewing.

I'm okay for this PR with this change, but please think about this to minimize the code change outside the focus of refactoring purpose.

RupinMittal committed:
Adopt more smart pointers in RemoteSourceBufferProxy
https://bugs.webkit.org/show_bug.cgi?id=279797
rdar://136114766

Reviewed by Ryosuke Niwa.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
(WebCore::SourceBufferPrivateClient::InitializationSegment::AudioTrackInformation::protectedTrack const):
(WebCore::SourceBufferPrivateClient::InitializationSegment::VideoTrackInformation::protectedTrack const):
(WebCore::SourceBufferPrivateClient::InitializationSegment::TextTrackInformation::protectedTrack const):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitializationSegment):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateHighestPresentationTimestampChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDurationChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateBufferedChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidDropSample):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveRenderingError):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateEvictionDataChanged):
(WebKit::RemoteSourceBufferProxy::append):
(WebKit::RemoteSourceBufferProxy::evictCodedFrames):
(WebKit::RemoteSourceBufferProxy::shutdown):

Canonical link: https://commits.webkit.org/283822@main
